cantusdb/turtle/csv2rdf_by_RDFLib.py

```diff
@@ -3,69 +3,63 @@
-    return ontology_dict
-    
 # a main operation function
-def convert_csv_to_turtle():
-    # TODO: maybe change this one to accomadate for multiple files
+def convert_csv_to_turtle(filename) -> Graph:
+    """
+    Adds all informations as RDF triples from the input filename into a graph and return it.
+    @Pre: input filename must be type string.
+    @Post: Returns a RDF.Graph that has all the triples
+    """
+    
     if len(sys.argv) != 2:
-        raise ValueError("Invalid number of input filename") 
+        raise ValueError("Invalid number of input filename")
     
-    g = Graph()
+    with open(filename, "r", encoding="utf-8") as csv_file:
+        # TODO: maybe change this one to accomadate for multiple files
 
-    csv_file = open(sys.argv[1], 'r', encoding='utf-8')
-    
-    csv_reader = csv.DictReader(csv_file)
-    ontology_dict = get_ontology_dict()
+        g = Graph()
 
-    # Convert each row to Turtle format and add it to the output
-    for row in csv_reader:
-        # the first column as the subject
-        absolute_uri = URIRef(row[list(row.keys())[0]])
-        g.add((absolute_uri, RDF.type, URIRef("http://www.wikidata.org/prop/direct/chant")))
+        csv_reader = csv.reader(csv_file)
+        with open("relations_mapping.json", "r") as mapper:
+            ontology_dict = json.load(mapper)
         
-        # extracting other informations
-        for k in row:
-            # if this key is the first column, skip
-            if k.__eq__(list(row.keys())[0]):
-                continue
-            
-            # finding the predicate from csv in the config dictionary, if not exit, skip
-            if k in list(ontology_dict.keys()):
-                predicate = URIRef(ontology_dict[k])
-            else:
-                continue
-            
-            # the object might be an URI or a literal
-            if validators.url(row[k]):
-                obj = URIRef(row[k])
+        header = next(csv_reader)
+
+        # Convert each row to Turtle format and add it to the output
+        for row in csv_reader:
+            # the first column as the subject
+            key_attribute = URIRef(row[0])
+            if "type" in list(ontology_dict.keys()):
+                g.add((key_attribute, RDF.type, URIRef(ontology_dict["type"])))
             else:
-                obj = Literal(row[k])
-            
-            # add this triple (absolute uri, predicate, object) to the graph
-            g.add((absolute_uri, predicate, obj))    
-        
-    csv_file.close()
-    
+                raise ValueError("No type specifications in the mapper file.")
+
+            # extracting other informations
+            # TODO: is the source description texts needed?
+            for i, element in enumerate(row[1:]):
+                predicate_cur = header[i]
+                # finding the predicate from csv in the config dictionary, if not exit, skip
+                if predicate_cur in list(ontology_dict.keys()):
```

Comment:
Consider reversing this so that you don't need the `else` --

```
if predicate not in ...: 
    continue

predicate = URIRef(...)
```

Reply:
Also, just looking at this more closely... 

`if predicate_cur in list(ontology_dict.keys())` is the same as `if predicate_cur in ontology_dict`, except for the fact that `x in list` is `O(n)`, while `x in dict` is `O(1)` (searching a list vs. searching a hashmap). It's not a big deal here because the lists are so small, but it's a good habit to always know the most performant path. 

Slow code that you don't write is code you don't have to optimize later.